src/backend/distributed/shardsplit/shardsplit_decoder.c

```diff
@@ -90,6 +90,40 @@ replication_origin_filter_cb(LogicalDecodingContext *ctx, RepOriginId origin_id)
 }
 
 
+/*
+ * update_replication_progress is copied from Postgres 15. We use it to send keepalive
+ * messages when we are filtering out the wal changes resulting from the initial copy.
+ * If we do not send out messages long enough, wal reciever will time out.
+ */
+static void
+update_replication_progress(LogicalDecodingContext *ctx, bool skipped_xact)
+{
+	static int changes_count = 0;
```

Comment:
should this declared as an "unsigned int"?
theoratically, this might overflow for a signed int, if the number of rows processed cross 2GB here..

Reply:
Given we cap the count at `CHANGES_THRESHOLD`, and only increment the variable in this function, I don't think there is any chance of overflowing.